### PR TITLE
switch UpdatePodSpecForObject to work on v1.PodSpec, use info.VersionedObject, and avoid conversion completely

### PR DESCRIPTION
**Pull request by zjj2wry**
**What this PR does / why we need it**:
ref #54212
ref #https://github.com/kubernetes/kubectl/issues/83

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
cc @liggitt @pwittrock 
**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by zjj2wry** [comment]
cc @liggitt @pwittrock

File: pkg/kubectl/cmd/util/factory_client_access.go [hunk 1]
```diff
@@ -235,25 +238,25 @@ func (f *ring0Factory) JSONEncoder() runtime.Encoder {
 	return legacyscheme.Codecs.LegacyCodec(legacyscheme.Registry.EnabledVersions()...)
 }
 
-func (f *ring0Factory) UpdatePodSpecForObject(obj runtime.Object, fn func(*api.PodSpec) error) (bool, error) {
```
Comment (zjj2wry) [inline pkg/kubectl/cmd/util/factory_client_access.go]: @liggitt UpdatePodSpecForObject should support all version, right?

File: pkg/kubectl/cmd/set/helper.go [hunk 2]
```diff
@@ -21,20 +21,20 @@ import (
 	"io"
 	"strings"
 
+	"k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/apimachinery/pkg/util/strategicpatch"
-	"k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/api/legacyscheme"
```
Comment (zjj2wry) [inline pkg/kubectl/cmd/set/helper.go]: @liggitt this pkg should alse remove? we want remove all `kubernetes/kubernetes pkg` from kubectl?

**Comment by zjj2wry** [comment]
@liggitt leave some questions for you, thanks :)

File: pkg/kubectl/cmd/set/helper.go [hunk 3]
```diff
@@ -139,7 +139,14 @@ func CalculatePatch(patch *Patch, encoder runtime.Encoder, mutateFn patchFn) boo
 		return false
 	}
 
-	// TODO: should be via New
+	// Create the versioned struct from the type defined in the restmapping
+	// (which is the API version we'll be submitting the patch to)
+	_, err := legacyscheme.Scheme.New(patch.Info.Mapping.GroupVersionKind)
```
Comment (liggitt) [inline pkg/kubectl/cmd/set/helper.go]: what is the purpose of this call? the result isn't used...

File: pkg/kubectl/cmd/set/helper.go [hunk 4]
```diff
@@ -139,7 +139,14 @@ func CalculatePatch(patch *Patch, encoder runtime.Encoder, mutateFn patchFn) boo
...
+	if err != nil {
+		patch.Err = err
+		return true
+	}
+
 	versioned, err := patch.Info.Mapping.ConvertToVersion(patch.Info.Object, patch.Info.Mapping.GroupVersionKind.GroupVersion())
```
Comment (liggitt) [inline pkg/kubectl/cmd/set/helper.go]: I'd actually be interested to see if `patch.Info.VersionedObject` could be used instead, removing the need to call ConvertToVersion here or to create versionedEncoder above

File: pkg/kubectl/cmd/set/set_image.go [hunk 5]
```diff
@@ -226,9 +226,8 @@ func (o *ImageOptions) Run() error {
 			return nil
 		})
 		if transformed && err == nil {
-			// TODO: switch UpdatePodSpecForObject to work on v1.PodSpec, use info.VersionedObject, and avoid conversion completely
 			versionedEncoder := legacyscheme.Codecs.EncoderForVersion(o.Encoder, info.Mapping.GroupVersionKind.GroupVersion())
```
Comment (liggitt) [inline pkg/kubectl/cmd/set/set_image.go]: can remove versionedEncoder and use o.Encoder directly if you're encoding an already versioned object (applies in all places)

File: pkg/kubectl/cmd/util/factory_client_access.go [hunk 6]
```diff
@@ -235,25 +238,25 @@ func (f *ring0Factory) JSONEncoder() runtime.Encoder {
...
 		return true, fn(&t.Spec)
-	case *api.ReplicationController:
+	case *v1.ReplicationController:
 		if t.Spec.Template == nil {
-			t.Spec.Template = &api.PodTemplateSpec{}
+			t.Spec.Template = &v1.PodTemplateSpec{}
 		}
 		return true, fn(&t.Spec.Template.Spec)
-	case *extensions.Deployment:
+	case *appsv1beta1.Deployment:
```
Comment (liggitt) [inline pkg/kubectl/cmd/util/factory_client_access.go]: deployment exists in extensions/v1beta1, apps/v1beta1, and apps/v1beta2... need to select all three

File: pkg/kubectl/cmd/util/factory_client_access.go [hunk 7]
```diff
@@ -235,25 +238,25 @@ func (f *ring0Factory) JSONEncoder() runtime.Encoder {
...
 		return true, fn(&t.Spec.Template.Spec)
-	case *extensions.DaemonSet:
+	case *appsv1beta2.DaemonSet:
```
Comment (liggitt) [inline pkg/kubectl/cmd/util/factory_client_access.go]: extensions/v1beta1, apps/v1beta2, apps/v1

File: pkg/kubectl/cmd/util/factory_client_access.go [hunk 8]
```diff
@@ -235,25 +238,25 @@ func (f *ring0Factory) JSONEncoder() runtime.Encoder {
...
 		return true, fn(&t.Spec.Template.Spec)
-	case *extensions.ReplicaSet:
+	case *appsv1beta2.ReplicaSet:
```
Comment (liggitt) [inline pkg/kubectl/cmd/util/factory_client_access.go]: extensions/v1beta1, apps/v1beta2

File: pkg/kubectl/cmd/util/factory_client_access.go [hunk 9]
```diff
@@ -235,25 +238,25 @@ func (f *ring0Factory) JSONEncoder() runtime.Encoder {
...
 		return true, fn(&t.Spec.Template.Spec)
-	case *apps.StatefulSet:
+	case *appsv1beta1.StatefulSet:
```
Comment (liggitt) [inline pkg/kubectl/cmd/util/factory_client_access.go]: apps/v1beta1, apps/v1beta2

File: pkg/kubectl/cmd/util/factory_client_access.go [hunk 10]
```diff
@@ -235,25 +238,25 @@ func (f *ring0Factory) JSONEncoder() runtime.Encoder {
...
 		return true, fn(&t.Spec.Template.Spec)
-	case *batch.Job:
+	case *batchv1.Job:
 		return true, fn(&t.Spec.Template.Spec)
 	default:
```
Comment (liggitt) [inline pkg/kubectl/cmd/util/factory_client_access.go]: also should probably support CronJob (batch/v1beta1, batch/v2alpha1)

File: pkg/kubectl/cmd/util/factory_client_access.go [hunk 11]
```diff
@@ -235,25 +238,25 @@ func (f *ring0Factory) JSONEncoder() runtime.Encoder {
 	return legacyscheme.Codecs.LegacyCodec(legacyscheme.Registry.EnabledVersions()...)
 }
 
-func (f *ring0Factory) UpdatePodSpecForObject(obj runtime.Object, fn func(*api.PodSpec) error) (bool, error) {
+func (f *ring0Factory) UpdatePodSpecForObject(obj runtime.Object, fn func(*v1.PodSpec) error) (bool, error) {
 	// TODO: replace with a swagger schema based approach (identify pod template via schema introspection)
 	switch t := obj.(type) {
```
Comment (liggitt) [inline pkg/kubectl/cmd/util/factory_client_access.go]: The number of objects and the number of versions reveals the unsustainability of this approach (especially as the workload API objects are getting promoted to new API versions - @kow3ns). This is ok for now, but you should add a reflective unit test that will fail if a compiled-in versioned type containing a pod spec is added and not handled here. Long-term, we need to find a way for kubectl to discover and extract pod spec fields from arbitrary objects that are not compiled in.

**Comment by zjj2wry** [comment]
@liggitt thank you very much, will update this tomorrow.
unassign @ghodss

File: pkg/kubectl/cmd/set/helper.go (hunk 3 above)
Comment (zjj2wry) [reply pkg/kubectl/cmd/set/helper.go]: Do your know this Todo mean?

**Comment by zjj2wry** [comment]
```
--- FAIL: TestServiceAccountRemote (0.00s)
        Error Trace:    set_serviceaccount_test.go:194
	Error:      	Received unexpected error:
	            	error: replicasets/nginx expected pointer, but got invalid kind
        Error Trace:    set_serviceaccount_test.go:194
	Error:      	Received unexpected error:
	            	error: daemonsets/nginx expected pointer, but got invalid kind
        Error Trace:    set_serviceaccount_test.go:194
	Error:      	Received unexpected error:
	            	error: replicationcontrollers/nginx expected pointer, but got invalid kind
        Error Trace:    set_serviceaccount_test.go:194
	Error:      	Received unexpected error:
	            	error: deployments/nginx expected pointer, but got invalid kind
        Error Trace:    set_serviceaccount_test.go:194
	Error:      	Received unexpected error:
	            	error: jobs/nginx expected pointer, but got invalid kind
        Error Trace:    set_serviceaccount_test.go:194
	Error:      	Received unexpected error:
	            	error: statefulsets/nginx expected pointer, but got invalid kind
```

File: pkg/kubectl/cmd/set/set_serviceaccount_test.go [hunk 12]
```diff
@@ -93,109 +86,6 @@ func TestServiceAccountLocal(t *testing.T) {
 	}
 }
 
-func TestServiceAccountRemote(t *testing.T) {
```
Comment (zjj2wry) [inline pkg/kubectl/cmd/set/set_serviceaccount_test.go]: can not get info.VersionedObject in this cases. @liggitt do you know this?

now just remove this for test others, will retrieve it~

**Comment by zjj2wry** [comment]
`set selector` and `set subject` not use `f.UpdatePodSpecForObject`, will create another PR fix it.

**Comment by zjj2wry** [comment]
--- FAIL: TestServiceAccountLocal (0.06s)
        Error Trace:    set_serviceaccount_test.go:84
	Error:      	Received unexpected error:
	            	error: statefulsets/minio converting (v1beta2.StatefulSet).ObservedGeneration to (v1beta1.StatefulSet).ObservedGeneration: type names don't match (int64, ), and no conversion 'func (int64, *int64) error' registered.

have sone test need fix,

File: pkg/kubectl/cmd/set/set_subject.go [hunk 13]
```diff
@@ -219,7 +219,6 @@ func (o *SubjectOptions) Run(f cmdutil.Factory, fn updateSubjects) error {
 
 		transformed, err := updateSubjectForObject(info.Object, subjects, fn)
 		if transformed && err == nil {
-			// TODO: switch UpdatePodSpecForObject to work on v1.PodSpec, use info.VersionedObject, and avoid conversion completely
 			versionedEncoder := legacyscheme.Codecs.EncoderForVersion(o.Encoder, info.Mapping.GroupVersionKind.GroupVersion())
```
Comment (liggitt) [inline pkg/kubectl/cmd/set/set_subject.go]: should be able to remove `versionedEncoder ...` and use o.Encoder directly

Comment (zjj2wry) [reply pkg/kubectl/cmd/set/set_subject.go]: `set_subject` and `set_selector` not use f.UpdatePodSpecForObject, will fix it when other test pass ci.

Comment (zjj2wry) [reply pkg/kubectl/cmd/set/set_subject.go]: when i use v1beta2.StatefulSet yaml will occur:
`
statefulsets/minio converting (v1beta2.StatefulSet).ObservedGeneration to (v1beta1.StatefulSet).ObservedGeneration: type names don't match (int64, ), and no conversion 'func (int64, *int64) error' registered.`

Comment (liggitt) [reply pkg/kubectl/cmd/set/set_subject.go]: this means you're trying to convert between two external versions directly and falling back to reflective conversion paths, which is very odd.

Comment (liggitt) [reply pkg/kubectl/cmd/set/set_subject.go]: `testapi.Default = testapi.Groups[input.apiGroup]` in the test is using the default test version for that group. I'm guessing that does not match the version of the API object in the test fixture

Comment (zjj2wry) [reply pkg/kubectl/cmd/set/set_subject.go]: when i use binary alse reproduce this:
```admindeMacBook-Pro-2:kubectl admin$ ./kubectl set image -f ../../examples/storage/minio/minio-distributed-statefulset.yaml *=gninxxx --local -o yaml
onverting (v1beta2.StatefulSet).ObservedGeneration to (v1beta1.StatefulSet).ObservedGeneration: type names don't match (int64, ), and no conversion 'func (int64, *int64) error' registered.```

**Comment by zjj2wry** [comment]
leaves two questiones:
1、can not set `v1beta2.StatefulSet`
--- FAIL: TestServiceAccountLocal (0.06s)
Error Trace: set_serviceaccount_test.go:84
Error: Received unexpected error:
error: statefulsets/minio converting (v1beta2.StatefulSet).ObservedGeneration to (v1beta1.StatefulSet).ObservedGeneration: type names don't match (int64, ), and no conversion 'func (int64, *int64) error' registered.

2、i can not get versionedObject from remote
--- FAIL: TestServiceAccountRemote (0.00s)
        Error Trace:    set_serviceaccount_test.go:194
	Error:      	Received unexpected error:
	            	error: replicasets/nginx expected pointer, but got invalid kind
        Error Trace:    set_serviceaccount_test.go:194
	Error:      	Received unexpected error:
	            	error: daemonsets/nginx expected pointer, but got invalid kind
        Error Trace:    set_serviceaccount_test.go:194
	Error:      	Received unexpected error:
	            	error: replicationcontrollers/nginx expected pointer, but got invalid kind
        Error Trace:    set_serviceaccount_test.go:194
	Error:      	Received unexpected error:
	            	error: deployments/nginx expected pointer, but got invalid kind
        Error Trace:    set_serviceaccount_test.go:194
	Error:      	Received unexpected error:
	            	error: jobs/nginx expected pointer, but got invalid kind
        Error Trace:    set_serviceaccount_test.go:194
	Error:      	Received unexpected error:
	            	error: statefulsets/nginx expected pointer, but got invalid kind

**Comment by liggitt** [comment]
please revert the change that removed TestServiceAccountRemote, and we can try to resolve the error

**Comment by zjj2wry** [comment]
@liggitt sorry for delay, have enough time debug it, and seems i block others... :(

**Comment by zjj2wry** [comment]
@liggitt i debug this, now just `StreamVisit` will have versionedObject. this is on purpose? 

i think we can convet info.Object assign to info.VersionedObject in `CalculatePatch` func. then needn't update `kubectl/resource` package.

**Comment by zjj2wry** [comment]
@liggitt i can use this yaml in my local now.
```
admindeMacBook-Pro-2:kubectl admin$ ./kubectl set image -f ../../examples/storage/minio/minio-distributed-statefulset.yaml *=nginx
statefulset "minio" image updated
```

File: pkg/kubectl/cmd/util/factory_builder.go [hunk 14]
```diff
@@ -130,7 +130,7 @@ func (f *ring2Factory) PrintObject(cmd *cobra.Command, isLocal bool, mapper meta
 	if err != nil {
 		return err
 	}
-
+	mapping.GroupVersionKind = gvks[0]
 	printer, err := f.PrinterForMapping(cmd, isLocal, nil, mapping, false)
```
Comment (zjj2wry) [inline pkg/kubectl/cmd/util/factory_builder.go]: @liggitt when use versionObject for printer. though my yaml is `v1beta2`, but mapping.GroupVersionKind  is `v1beta1`. add this line will fix this.
```
admindeMacBook-Pro-2:kubectl admin$ ./kubectl set image -f ../../examples/storage/minio/minio-distributed-statefulset.yaml *=test --local -o yaml
apiVersion: apps/v1beta2
kind: StatefulSet
metadata:
  creationTimestamp: null
  name: minio
spec:
  replicas: 4
...

```

Comment (zjj2wry) [reply pkg/kubectl/cmd/util/factory_builder.go]: and last version kubectl, will update apiVersion to `v1beta1` though my yaml version is `v1beta2`:
```
admindeMacBook-Pro-2:kubernetes admin$ kubectl set image -f examples/storage/minio/minio-distributed-statefulset.yaml *=test --local -o yaml
apiVersion: apps/v1beta1
kind: StatefulSet
metadata:
  creationTimestamp: null
  name: minio
  namespace: default

```

File: pkg/kubectl/cmd/util/factory_builder.go [hunk 15]
```diff
@@ -130,7 +130,7 @@ func (f *ring2Factory) PrintObject(cmd *cobra.Command, isLocal bool, mapper meta
 	if err != nil {
 		return err
 	}
-
+	mapping.GroupVersionKind = gvks[0]
```
Comment (liggitt) [inline pkg/kubectl/cmd/util/factory_builder.go]: revert this. the printing code is not prepared to use external objects yet

**Comment by liggitt** [comment]
all the places we pass an object to local printers, it must be an internal object, because printers still require internal objects. making printers not require internal versions is desired, but is a way bigger change.

set_env.go:
```
if err := o.PrintObject(o.Cmd, o.Local, o.Mapper, info.Object, o.Out); err != nil {
```

set_resources.go:
```
if err := o.PrintObject(o.Cmd, o.Local, o.Mapper, info.Object, o.Out); err != nil {
```

set_serviceaccount.go
```
saConfig.saPrint(patch.Info.Object)
```

if we're dealing with creating patches against the versioned object, that means we need to convert that resulting versioned object to internal just before printing locally. for set_serviceaccount, this could look like this:
```diff
diff --git a/pkg/kubectl/cmd/set/set_serviceaccount.go b/pkg/kubectl/cmd/set/set_serviceaccount.go
index 5914445d66..7ef0cba89e 100644
--- a/pkg/kubectl/cmd/set/set_serviceaccount.go
+++ b/pkg/kubectl/cmd/set/set_serviceaccount.go
@@ -58,6 +58,7 @@ type serviceAccountConfig struct {
        fileNameOptions        resource.FilenameOptions
        mapper                 meta.RESTMapper
        encoder                runtime.Encoder
+       decoder                runtime.Decoder
        out                    io.Writer
        err                    io.Writer
        dryRun                 bool
@@ -107,6 +108,7 @@ func NewCmdServiceAccount(f cmdutil.Factory, out, err io.Writer) *cobra.Command
 func (saConfig *serviceAccountConfig) Complete(f cmdutil.Factory, cmd *cobra.Command, args []string) error {
        saConfig.mapper, _ = f.Object()
        saConfig.encoder = f.JSONEncoder()
+       saConfig.decoder = f.Decoder(true)
        saConfig.shortOutput = cmdutil.GetFlagString(cmd, "output") == "name"
        saConfig.record = cmdutil.GetRecordFlag(cmd)
        saConfig.changeCause = f.Command(cmd, false)
@@ -162,7 +164,20 @@ func (saConfig *serviceAccountConfig) Run() error {
                        continue
                }
                if saConfig.local || saConfig.dryRun {
-                       saConfig.saPrint(patch.Info.Object)
+                       // TODO: remove once printers can deal with versioned objects
+                       // convert back to internal for printing
+                       versionedData, err := runtime.Encode(saConfig.encoder, info.VersionedObject)
+                       if err != nil {
+                               patchErrs = append(patchErrs, fmt.Errorf("error: %s/%s %v\n", info.Mapping.Resource, info.Name, err))
+                               continue
+                       }
+                       internalObj, err := runtime.Decode(saConfig.decoder, versionedData)
+                       if err != nil {
+                               patchErrs = append(patchErrs, fmt.Errorf("error: %s/%s %v\n", info.Mapping.Resource, info.Name, err))
+                               continue
+                       }
+                       saConfig.saPrint(internalObj)
                        continue
                }
                patched, err := resource.NewHelper(info.Client, info.Mapping).Patch(info.Namespace, info.Name, types.StrategicMergePatchType, patch.Patch)
```

File: pkg/kubectl/cmd/util/factory_builder.go [hunk 16]
```diff
@@ -127,6 +127,7 @@ func (f *ring2Factory) PrintObject(cmd *cobra.Command, isLocal bool, mapper meta
 	}
 
 	mapping, err := mapper.RESTMapping(gvks[0].GroupKind())
+	fmt.Println(gvks[0].GroupKind(), gvks[0].GroupVersion().Version, gvks)
```
Comment (zjj2wry) [inline pkg/kubectl/cmd/util/factory_builder.go]: @liggitt this is a bug in master and just reproduce in daemonset...
```
admindeMacBook-Pro-2:kubectl admin$ ./kubectl set image -f ../../test/fixtures/doc-yaml/admin/daemon.yaml *=nginx --local -o yaml
no matches for apps/, Kind=DaemonSet
apps/, Kind=DaemonSet matches multiple kinds [apps/v1beta2, Kind=DaemonSet apps/v1, Kind=DaemonSet]
```

**Comment by liggitt** [comment]
tracked down the `error: statefulsets/minio converting (v1beta2.StatefulSet).ObservedGeneration to (v1beta1.StatefulSet).ObservedGeneration: type names don't match (int64, ), and no conversion 'func (int64, *int64) error' registered.` error to the testapi Codec only knowing about a single version for each API group, and choosing a different one than is in the test fixture

at this point, the testapi factory is so different from the real factory the command runs with that it is questionable whether it is adding value... there are bugs and inflexibility in the testapi factory that do not exist in real life

File: pkg/kubectl/cmd/util/factory_client_access.go (hunk 11 above)
Comment (liggitt) [reply pkg/kubectl/cmd/util/factory_client_access.go]: missing appsv1.Deployment, appsv1.StatefulSet, appsv1beta1.ReplicaSet, appsv1.ReplicaSet

Comment (liggitt) [reply pkg/kubectl/cmd/util/factory_client_access.go]: this is constantly going to drift as new versions of types are added

**Comment by zjj2wry** [comment]
@liggitt thank you, it is real helpful : )

**Comment by zjj2wry** [comment]
Fixs #55407 #55408

**Comment by zjj2wry** [comment]
@liggitt this PR ready, please review again, thanks

File: pkg/kubectl/cmd/set/set_resources.go [hunk 17]
```diff
@@ -128,7 +127,7 @@ func NewCmdResources(f cmdutil.Factory, out io.Writer, errOut io.Writer) *cobra.
 }
 
 func (o *ResourcesOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, args []string) error {
-	o.Mapper, o.Typer = f.Object()
+	o.Mapper, _ = f.Object()
```
Comment (liggitt) [inline pkg/kubectl/cmd/set/set_resources.go]: why this change?

Comment (zjj2wry) [reply pkg/kubectl/cmd/set/set_resources.go]: this Typer not in use, so i clean it.

**Comment by zjj2wry** [comment]
@liggitt have clean all not in use o.Typer, ptal

**Comment by liggitt** [comment]
seeing this when running on a v1beta1 statefulset:

```
$ kubectl set serviceaccount -f ss.v1beta1.yaml foo -o yaml
error: converting (v1beta1.StatefulSet).ObservedGeneration to (v1.StatefulSet).ObservedGeneration: type names don't match (, int64), and no conversion 'func (*int64, int64) error' registered.

$ kubectl set serviceaccount -f ss.v1beta1.yaml foo -o yaml --local
error: converting (v1beta1.StatefulSet).ObservedGeneration to (v1.StatefulSet).ObservedGeneration: type names don't match (, int64), and no conversion 'func (*int64, int64) error' registered.
```

on master it works properly

**Comment by liggitt** [comment]
it looks like the printer is not preferring the existing external version if specified. This change in `func (f *ring2Factory) PrintObject` makes it work as expected:

```go
	// Prefer the existing external version if specified
	var preferredVersion []string
	if gvks[0].Version != "" && gvks[0].Version != runtime.APIVersionInternal {
		preferredVersion = []string{gvks[0].Version}
	}

	mapping, err := mapper.RESTMapping(gvks[0].GroupKind(), preferredVersion...)
	if err != nil {
		return err
	}
```

File: pkg/kubectl/cmd/util/factory_client_access.go (hunk 11 above)
Comment (zjj2wry) [reply pkg/kubectl/cmd/util/factory_client_access.go]: thanks you ～ add it now. is it possible can have accessor return v1.podtemplate?
```
type ObjectMetaAccessor interface {
	GetObjectMeta() Object
}

// Object lets you work with object metadata from any of the versioned or
// internal API objects. Attempting to set or retrieve a field on an object that does
// not support that field (Name, UID, Namespace on lists) will be a no-op and return
// a default value.
type Object interface {
	GetNamespace() string
	SetNamespace(namespace string)
	GetName() string
	SetName(name string)
	GetGenerateName() string
	SetGenerateName(name string)
	GetUID() types.UID
	SetUID(uid types.UID)
	GetResourceVersion() string
	SetResourceVersion(version string)
	GetGeneration() int64
	SetGeneration(generation int64)
	GetSelfLink() string
	SetSelfLink(selfLink string)
	GetCreationTimestamp() Time
	SetCreationTimestamp(timestamp Time)
	GetDeletionTimestamp() *Time
	SetDeletionTimestamp(timestamp *Time)
	GetDeletionGracePeriodSeconds() *int64
	SetDeletionGracePeriodSeconds(*int64)
	GetLabels() map[string]string
	SetLabels(labels map[string]string)
	GetAnnotations() map[string]string
	SetAnnotations(annotations map[string]string)
	GetInitializers() *Initializers
	SetInitializers(initializers *Initializers)
	GetFinalizers() []string
	SetFinalizers(finalizers []string)
	GetOwnerReferences() []OwnerReference
	SetOwnerReferences([]OwnerReference)
	GetClusterName() string
	SetClusterName(clusterName string)
}
```

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/54554/files#r149966979 is still outstanding as well

**Comment by zjj2wry** [comment]
@liggitt add `preferredVersion` will can work properly.

File: pkg/kubectl/cmd/util/factory_client_access.go (hunk 11 above)
Comment (liggitt) [reply pkg/kubectl/cmd/util/factory_client_access.go]: > is it possible can have accessor return v1.podtemplate?

no, ObjectMetaAccessor only returns things that are part of metadata that all objects have

**Comment by zjj2wry** [comment]
@liggitt thanks you help debug for this PR. 🎁

**Comment by liggitt** [comment]
thanks, go ahead and squash down to a single commit

**Comment by zjj2wry** [comment]
@liggitt ready go, have squash, ptal

**Comment by smarterclayton** [comment]
This is not aligned with #55660, I may revert this.

**Comment by zjj2wry** [comment]
got～ it is big enhance

**Comment by smarterclayton** [comment]
Take a look at the last commit in #55660 - I think with the removal of versioned object it's clear what the desired outcome is in those cases.
